File: Src/IronJS/Caches.fs
namespace IronJS.Caches

open System
open IronJS.Support.Aliases
open IronJS.SplayTree

///
type LimitCache<'k, 'v when 'k : equality>(size:int) =
  let storage = ref List.empty<'k * 'v>

  member x.Lookup(key:'k, value:Lazy<'v>) =
    match !storage |> List.tryFind (fun (k, _) -> k = key) with
    | None ->
      storage := (key, value.Value) :: !storage

      if (!storage).Length > size * 2 then
        storage := !storage |> Seq.take size |> Seq.toList

      value.Value

    | Some(_, cached) -> 
      cached

///
type SplayCache<'k, 'v when 'k :> IComparable<'k> and 'v : equality>() =
  let storage = new SplayTree<'k, 'v>()

  member x.Lookup(key:'k, value:Lazy<'v>) =
      let result = storage.TryGetValue(key)
      if result.IsNone then
        let result = value.Value
        storage.[key] <- result
        if storage.Count > 1024 then
          storage.Trim(10)
        result
      else
         result.Value

///
type WeakCache<'TKey, 'TValue when 'TKey : equality>() =
  let cache = new System.Collections.Generic.Dictionary<'TKey, WeakReference>()

  member x.Lookup (key:'TKey, create:Lazy<'TValue>) =
    let cacheHit, reference = cache.TryGetValue(key)
    if cacheHit then
      match reference.Target with
      | :? 'TValue as value -> value
      | _ ->
        let value = create.Value
        let reference = new WeakReference(value)
        cache.[key] <- reference
        value
    else
      let value = create.Value
      let reference = new WeakReference(value)
      cache.Add(key, reference)
      value

File: Src/IronJS.Runtime/Caches.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace IronJS.Runtime.Caches
{
    public class LimitCache<K, V>
        where K : IEquatable<K>
        where V : class
    {
        int size;
        List<Tuple<K, V>> storage =
            new List<Tuple<K, V>>();

        public LimitCache(int halfSize)
        {
            this.size = halfSize * 2;
        }

        public V Lookup(K key, Lazy<V> value)
        {
            var result = storage.FirstOrDefault(x => x.Item1.Equals(key));

            if (result == null)
            {
                storage.Add(Tuple.Create(key, value.Value));

                if (storage.Count > size)
                {
                    storage = storage.Take(size).ToList();
                }
            }

            return result.Item2;
        }
    }

    public class SplayCache<K, V>
        where K : IComparable<K>
    {
        SplayTree<K, V> storage =
            new SplayTree<K, V>();

        public V Lookup(K key, Lazy<V> value)
        {
            V cached;

            if (!storage.TryGetValue(key, out cached))
            {
                storage[key] = cached = value.Value;

                if (storage.Count > 1024)
                {
                    storage.Trim(10);
                }
            }

            return cached;
        }
    }

    public class WeakCache<K, V>
    {
        Dictionary<K, WeakReference> storage =
            new Dictionary<K, WeakReference>();

        public V Lookup(K key, Func<V> value)
        {
            WeakReference cached;

            if (storage.TryGetValue(key, out cached) && cached.Target is V)
            {
                return (V)cached.Target;
            }

            var newValue = value();
            storage[key] = new WeakReference(newValue);
            return newValue;
        }
    }
}
